File: glut_opengl_estudos/06_double_buffer_fps.py
"""
Aula 06 — Double buffering e medição simples de FPS

Conceitos:
- `GLUT_DOUBLE`: renderiza em back buffer e troca com `glutSwapBuffers`.
- `glutIdleFunc`: atualiza continuamente a cena.
- FPS médio por janela temporal.

Como executar:
    python3 06_double_buffer_fps.py
"""

from time import perf_counter

from OpenGL.GL import (
    GL_COLOR_BUFFER_BIT,
    GL_QUADS,
    glBegin,
    glClear,
    glClearColor,
    glColor3f,
    glEnd,
    glLoadIdentity,
    glRotatef,
    glVertex2f,
)
from OpenGL.GLUT import GLUT_DOUBLE, GLUT_RGB, glutCreateWindow, glutDisplayFunc, glutIdleFunc, glutInit, glutInitDisplayMode, glutInitWindowSize, glutMainLoop, glutPostRedisplay, glutSwapBuffers

angle = 0.0
frame_count = 0
start_time = perf_counter()


def draw_rotating_quad() -> None:
    """Desenha um quadrado rotacionando ao redor do centro."""
    glBegin(GL_QUADS)
    glColor3f(1.0, 0.6, 0.2)
    glVertex2f(-0.3, -0.3)
    glVertex2f(0.3, -0.3)
    glColor3f(0.2, 0.8, 1.0)
    glVertex2f(0.3, 0.3)
    glVertex2f(-0.3, 0.3)
    glEnd()


def display() -> None:
    """Renderiza um frame e atualiza medição de FPS."""
    global frame_count, start_time

    glClear(GL_COLOR_BUFFER_BIT)
    glLoadIdentity()
    glRotatef(angle, 0.0, 0.0, 1.0)
    draw_rotating_quad()
    glutSwapBuffers()

    frame_count += 1
    elapsed = perf_counter() - start_time
    if elapsed >= 1.0:
        fps = frame_count / elapsed
        print(f"FPS médio: {fps:.1f}")
        frame_count = 0
        start_time = perf_counter()


def idle() -> None:
    """Atualiza ângulo e solicita novo frame."""
    global angle
    angle = (angle + 0.35) % 360.0
    glutPostRedisplay()


def main() -> None:
    """Entry point da aula 06."""
    glutInit()
    glutInitDisplayMode(GLUT_DOUBLE | GLUT_RGB)
    glutInitWindowSize(900, 600)
    glutCreateWindow(b"Aula 06 - Double Buffer e FPS")
    glClearColor(0.04, 0.04, 0.06, 1.0)

    glutDisplayFunc(display)
    glutIdleFunc(idle)
    glutMainLoop()


if __name__ == "__main__":
    main()

File: glut_opengl_estudos/06_double_buffer_fps.cpp
/*
Aula 06 — Double buffering e medição simples de FPS (C++)

Conceitos:
- `GLUT_DOUBLE`: desenha no back buffer e troca com `glutSwapBuffers`.
- `glutIdleFunc`: animação contínua sem timer explícito.
- Janela temporal para cálculo de FPS médio.
*/

#include <GL/glut.h>
#include <chrono>
#include <iostream>
#include "cpp_glut_utils.hpp"

namespace {

float angleDeg = 0.0f;
int frameCount = 0;
auto startTime = std::chrono::steady_clock::now();

void drawRotatingQuad() {
    glBegin(GL_QUADS);
    glColor3f(1.0f, 0.6f, 0.2f); glVertex2f(-0.3f, -0.3f);
    glColor3f(1.0f, 0.6f, 0.2f); glVertex2f(0.3f, -0.3f);
    glColor3f(0.2f, 0.8f, 1.0f); glVertex2f(0.3f, 0.3f);
    glColor3f(0.2f, 0.8f, 1.0f); glVertex2f(-0.3f, 0.3f);
    glEnd();
}

void display() {
    glClear(GL_COLOR_BUFFER_BIT);
    glLoadIdentity();
    glRotatef(angleDeg, 0.0f, 0.0f, 1.0f);
    drawRotatingQuad();
    glutSwapBuffers();

    frameCount++;
    const auto now = std::chrono::steady_clock::now();
    const std::chrono::duration<double> elapsed = now - startTime;
    if (elapsed.count() >= 1.0) {
        const double fps = static_cast<double>(frameCount) / elapsed.count();
        std::cout << "FPS medio: " << fps << '\n';
        frameCount = 0;
        startTime = now;
    }
}

void idle() {
    angleDeg += 0.35f;
    if (angleDeg >= 360.0f) angleDeg -= 360.0f;
    glutPostRedisplay();
}

} // namespace

int main(int argc, char** argv) {
    glutInit(&argc, argv);
    aula::configureCloseBehavior();
    glutInitDisplayMode(GLUT_DOUBLE | GLUT_RGB);
    glutInitWindowSize(900, 600);
    glutCreateWindow("Aula 06 - Double Buffer e FPS (C++)");

    glClearColor(0.04f, 0.04f, 0.06f, 1.0f);
    glutDisplayFunc(display);
    glutIdleFunc(idle);
    glutMainLoop();
    return 0;
}
